### ISSUE TITLE
Custom headers

### ISSUE THREAD

**Issue by Ilaiwi** [namespace-ee/react-calendar-timeline#349]
**Is your feature request related to a problem? Please describe.**
Support adding custom headers inspired by #339 to support for example adding daytime intervals something like this:
![rtc-customheaders](https://user-images.githubusercontent.com/10817089/41958706-8948176a-79f3-11e8-81a0-4cb4fbe59e31.PNG)

This will make it possible also to solve #140 and #130 in clear way.

**Describe the solution you'd like**
This is inspired by the custom markers here #339. I think its API is very declarative and flexible
```jsx
<Timeline>
  <TimelineHeaders>
    <DateHeader varient="month" />
    <DateHeader varient="hour" />
    <CustomHeader>
      {/* custom renderer for this marker */}
      {({ styles, timlineContext, ... }) => {
        const days= canvasTimeEnd-timecontext.canvasTimeStart ...
        return {days.map(day => <div style={style} onClick={someCustomHandler}>
                        <div>1</div>
                        <div>2</div>
                        <div>3</div>
                    </div>)}
      }}
    </CustomHeader>
  </TimelineHeader>
</Timeline>
```

**Describe alternatives you've considered**
this might be done using plugins but still, you need to increase the length of the header to accommodate the needed space

**Additional context**
![rtc-customheaders](https://user-images.githubusercontent.com/10817089/41958706-8948176a-79f3-11e8-81a0-4cb4fbe59e31.PNG)

**Comment by mcMickJuice** [comment]
Thanks for this @Ilaiwi .  I agree that the API should work this way.  Here are a couple of terms so we're on the same page (please feel free to recommend different names for these)

* `PrimaryUnit` - the top unit. This unit is always "larger" than the `SecondaryUnit` and appears above the secondary unit.  In your screenshot, "Month" is the `PrimaryUnit`
* `SecondaryUnit` - the unit below the `PrimaryUnit`.  This unit maps directly to the columns appearing in the timeline. I.e. The level of granularity you're viewing your timeline items.
* `CustomUnit` - this could be a tertiary row in the header where the user could specify custom units. This `CustomUnit` is not yet implemented and wouldn't be a part of this initial implementation.
* `HeaderUnitRow` - row that contains header units
* `HeaderUnitLabel` - container (element) for a header unit

A few things about this header:

* height of the header overall and height of each `HeaderUnitRow` is flexible (previously it was not)
* width of individual header units is *not* flexible.  This makes sense as we need to keep the positioning and width of `HeaderUnitLabel`s in line with the timeline columns. That said, if user has control of rendering, they can control fontsize scale based on width of the `HeaderUnitLabel`.  Or they can wrap the text...or truncate it.
* click handlers can be attached to these elements - https://github.com/namespace-ee/react-calendar-timeline/issues/140

What else do we need to think about for this feature?  What else would this feature help accommodate?

**Comment by Ilaiwi** [comment]
Hi @mcMickJuice, I started working on this in #391 and I would love to take your feedback on the terminology I used so far. 

So a default header would look like this 
```
<TimelineHeaders>
        <SidebarHeader />
        <DateHeader primaryHeader />
        <DateHeader />
</TimelineHeaders>
```
So I have 4 components 
* `TimelineHeaders` which is the renderer for all the header section above the body of the timeline
*  `SidebarHeader` which renders either the `right` or `left` sidebar headers. It uses child function to render the header and has a prop called `variant`.
* `CustomHeader` which is the base for all header placed above the scrollable calendar part. it accepts `unit` which could be `day`, `hour` ..... . Also the component uses a child function to render the header. It passes `rootProvidedProps`, `intervals` which is an array of time intervals with `providedProps`. This component is very basic and is used as a base for the other headers or for very customized headers
* `DateHeader` which represents the header above the scrollable calendar part. This is Based on the `CustomHeader` component and is built on top of it. The component accepts the following props:
    - `primaryHeader` which will make the header unit and style the main header 
    - `secondaryHeader` which will make the unit the same as the timeline unit and has the "sub header" style.
   - `unit` same as `CustomHeader` and will override the above props.
   - `labelFormat`: could be an object, string or function to format a header interval.
   - child function renderer: optional prop to render the intervals

Most of the features above are implemented or will have these features when I am done. 

I am also thinking of adding the ability passing the `unit` as an interval of time in milliseconds or as an `moment` interval.

I am thinking also of generalizing the calculation of intervals in the headers and calendar part.

** Please note that this component will be converted to use prop getters for consistency 

Please let me know what do you think.

**Comment by danielmorine** [comment]
How can I control the state change of the datetime in this.state? I want to control it with a date picker

**Comment by Ilaiwi** [comment]
released in 0.25.0
